library/core/src/alloc/global.rs

```diff
@@ -53,6 +53,25 @@ use crate::ptr;
 /// * `Layout` queries and calculations in general must be correct. Callers of
 ///   this trait are allowed to rely on the contracts defined on each method,
 ///   and implementors must ensure such contracts remain true.
+///
+/// * You may not rely on allocations actually happening, even if there are explicit
+///   heap allocations in the source. The optimizer may detect allocation/deallocation
+///   pairs that it can instead move to stack allocations/deallocations and thus never
+///   invoke the allocator here.
+///   More concretely, the following code example is unsound, irrespective of whether your
+///   custom allocator allows counting how many allocations have happened.
+///
+///   ```text
+///   drop(Box::new(42));
+///   let number_of_heap_allocs = /* call private allocator API */;
+///   unsafe { std::intrinsics::assume(number_of_heap_allocs > 0); }
+///   ```
+///
+///   Note that allocation/deallocation pairs being moved to the stack is not the only
+///   optimization that can be applied. You may generally not rely on heap allocations
+///   happening, if they can be removed without changing program behaviour.
```

Comment:
May you add, that it can even be optimized out, if the allocator has side effects like printing to the terminal or counting a global variable.